### PR TITLE
ECF-RP-240: Update school import

### PR DESCRIPTION
**Pull request by ethanmills**
### Context
We now import school data from the private GIAS API, not the public file

### Changes proposed in this pull request
- Create a GIAS API client
- Update the import to use the custom GIAS extract

### Guidance to review
commit-wise

### Testing
- unit tests for import
- unit tests for api client

### CODE REVIEW

**Comment by ethanmills** [comment]
[Update school data import to use GIAS extract](https://trello.com/c/I0eihe4z/240-update-school-data-import-to-use-gias-extract)

File: lib/gias_api_client.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    parts.second.body.to_s
+  end
+
+  def temp_files_from_zipped_response(data)
+    files = {}
+
+    # We need to write the zip file to disk rather than read it from a buffer because of this:
+    # https://github.com/rubyzip/rubyzip#notice-about-zipinputstream
+    zipfile = Tempfile.new(encoding: "ASCII-8BIT")
+    zipfile.write(data.force_encoding("ASCII-8BIT"))
```
Comment (amazimbe) [inline lib/gias_api_client.rb:31]: I doubt we need this line if we have already specified the encoding in zipfile = Tempfile.new(encoding: "ASCII-8BIT")

**Review by amazimbe — APPROVED** [review]
Just one minor comment / question. Otherwise, all looks good.

File: Dockerfile [hunk 2]
```diff
@@ -47,7 +47,6 @@ FROM ${BASE_RUBY_IMAGE_WITH_GEMS_AND_NODE_MODULES} AS assets-precompile
 
 ENV GOVUK_APP_DOMAIN="http://localhost:3000" \
     GOVUK_WEBSITE_ROOT="http://localhost:3000" \
-    GIAS_API_ROOT="https://ea-edubase-api-prod.azurewebsites.net" \
```
Comment (kicferk1) [inline Dockerfile:50]: Do we not want to add the new env variables here?

File: app/services/school_data_importer.rb [hunk 3]
```diff
@@ -12,9 +12,10 @@ def initialize(logger, start_year = Time.zone.now.year)
     @start_year = start_year
   end
 
-  # TODO: Register and Partner 147: Figure out how exactly we want to handle school updating
   def run
-    CSV.foreach(schools_data_file.path, headers: true, encoding: "ISO-8859-1:UTF-8") do |row|
+    files = school_data_files
+
+    CSV.foreach(files["ecf_tech.csv"].path, headers: true, encoding: "ISO-8859-1:UTF-8") do |row|
```
Comment (kicferk1) [inline app/services/school_data_importer.rb:18]: Are there other files in there? What's in them?

File: app/services/school_data_importer.rb [hunk 4]
```diff
@@ -26,40 +27,29 @@ def run
...
     school.school_type_name = row.fetch("TypeOfEstablishment (name)")
-    school.capacity = row.fetch("SchoolCapacity").to_i
     school.address_line1 = row.fetch("Street")
     school.address_line2 = row.fetch("Locality")
     school.address_line3 = row.fetch("Town")
-    school.address_line4 = row.fetch("County (name)")
-    school.country = row.fetch("Country (name)")
     school.postcode = row.fetch("Postcode")
     school.ukprn = row.fetch("UKPRN")
-    school.previous_school_urn = row.fetch("PreviousEstablishmentNumber")
```
Comment (kicferk1) [inline app/services/school_data_importer.rb:51]: oof, scary

without this data, are we gonna have troubles figuring out which school is which after a change?

File: lib/gias_api_client.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+  def fetch_extract
+    client = Savon.client(wsdl: Rails.application.config.gias_api_schema)
+    res = client.call(
+      :get_extract,
+      message: { "tns:Id" => Rails.application.config.gias_extract_id },
+      wsse_auth: [Rails.application.config.gias_api_user, Rails.application.config.gias_api_password],
+      wsse_timestamp: true,
+    )
+    parts = parse_body(res.http)
```
Comment (kicferk1) [inline lib/gias_api_client.rb:21]: is our request using `http`? or `https`?

Not that I know what the hell `res.http` does, just wondering.

File: lib/gias_api_client.rb [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    parts.second.body.to_s
```
Comment (kicferk1) [inline lib/gias_api_client.rb:22]: Is this how we get rid of the request headers?

File: lib/gias_api_client.rb [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    end
+
+    files
+  end
+
+  def parse_body(http)
+    Mail::Part.new(
+      headers: http.headers,
+      body: http.body,
+    ).body.split!(boundary(http)).parts
```
Comment (kicferk1) [inline lib/gias_api_client.rb:50]: Are we re-purposing an email parser to deal with out zip file? :O

File: Dockerfile (hunk 2 above)
Comment (ethanmills) [reply Dockerfile:50]: Nope, we don't need this var for the build

File: app/services/school_data_importer.rb (hunk 3 above)
Comment (ethanmills) [reply app/services/school_data_importer.rb:18]: Yes. There's the groups file, the group links file, and the links file. Contents per https://www.get-information-schools.service.gov.uk/Downloads. Will be using the groups file later to set up MATs

File: app/services/school_data_importer.rb (hunk 4 above)
Comment (ethanmills) [reply app/services/school_data_importer.rb:51]: That's what the links file is for 🙂

File: lib/gias_api_client.rb (hunk 5 above)
Comment (ethanmills) [reply lib/gias_api_client.rb:21]: It's not that. The initial request will be https to https://ea-edubase-api-prod.azurewebsites.net/edubase/schema/service.wsdl
, and subsequent requests dictated by the contents of that schema. This just return the HTTPI object that has header and body information

File: lib/gias_api_client.rb (hunk 6 above)
Comment (ethanmills) [reply lib/gias_api_client.rb:22]: Yes, in essence. I adapted some code from a long dead library: [savon-multipart](https://github.com/savonrb/savon-multipart/blob/5e5a247f3fc5b5e2b25e2eaa76893fa34bda8adf/lib/savon/multipart/response.rb)

File: lib/gias_api_client.rb (hunk 7 above)
Comment (ethanmills) [reply lib/gias_api_client.rb:50]: See previous comment. It's basically a MIME parser, and the multipart response is most commonly found in emails

Comment (kicferk1) [reply lib/gias_api_client.rb:50]: Ech, well, not like I can think of a better way.

**Review by kicferk1 — APPROVED** [review]

File: lib/gias_api_client.rb (hunk 1 above)
Comment (ethanmills) [reply lib/gias_api_client.rb:31]: This is incredibly touchy about encodings. If you're happy with it, I'd like to leave this as it is, as it definitely works like this

**Pull request merged**